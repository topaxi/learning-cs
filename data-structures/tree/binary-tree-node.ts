import { HashMap } from '../hash/hash-map'

export class BinaryTreeNode<T, M = any> {
  readonly meta = new HashMap<M>()

  private _left: this | null = null
  private _right: this | null = null
  private _parent: this | null = null

  constructor(
    public value: T,
    protected readonly compare: (a: T, b: T) => number = (a: any, b: any) =>
      a - b
  ) {}

  get parent(): this | null {
    return this._parent
  }

  set left(node: this | null) {
    if (this._left !== null) {
      this._left._parent = null
    }

    this._left = node

    if (node !== null) {
      node._parent = this
    }
  }

  get left(): this | null {
    return this._left
  }

  set right(node: this | null) {
    if (this._right !== null) {
      this._right._parent = null
    }

    this._right = node

    if (node !== null) {
      node._parent = this
    }
  }

  get right(): this | null {
    return this._right
  }

  get leftHeight(): number {
    return this._left === null ? 0 : this._left.height + 1
  }

  get rightHeight(): number {
    return this._right === null ? 0 : this._right.height + 1
  }

  get height(): number {
    return Math.max(this.leftHeight, this.rightHeight)
  }

  get balanceFactor(): number {
    return this.leftHeight - this.rightHeight
  }

  removeChild(value: T): boolean {
    if (this._left !== null && this.compare(value, this._left.value) === 0) {
      this._left._parent = null
      this._left = null

      return true
    }

    if (this._right !== null && this.compare(value, this._right.value) === 0) {
      this._right._parent = null
      this._right = null

      return true
    }

    return false
  }

  removeChildNode(node: BinaryTreeNode<T, M>): boolean {
    return this.removeChild(node.value)
  }

  replaceChildNode(toReplace: this, replacement: this): boolean {
    if (
      this._left !== null &&
      this.compare(toReplace.value, this._left.value) === 0
    ) {
      this._left = replacement

      return true
    }

    if (
      this._right !== null &&
      this.compare(toReplace.value, this._right.value) === 0
    ) {
      this._right = replacement

      return true
    }

    return false
  }

  *[Symbol.iterator](): IterableIterator<T> {
    if (this.left !== null) yield* this.left
    yield this.value
    if (this.right !== null) yield* this.right
  }

  traverse(callback: (value: T) => any): void {
    for (let value of this) callback(value)
  }
}
