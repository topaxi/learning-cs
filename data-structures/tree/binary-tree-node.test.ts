import { BinaryTreeNode } from './binary-tree-node'

describe('BinaryTreeNode<T>', () => {
  describe('#toJSON()', () => {
    test('should serialize to JSON', () => {
      let node = new BinaryTreeNode('root')
      node.left = new BinaryTreeNode('left')
      node.right = new BinaryTreeNode('right')
      node.right.right = new BinaryTreeNode('rightright')
      node.right.left = new BinaryTreeNode('rightleft')

      expect(node).toMatchSnapshot()
    })
  })

  describe('#height', () => {
    test('should return tree height', () => {
      let node = new BinaryTreeNode('root')
      node.left = new BinaryTreeNode('left')
      node.right = new BinaryTreeNode('right')
      node.right.right = new BinaryTreeNode('rightright')
      node.right.left = new BinaryTreeNode('rightleft')

      expect(node.height).toBe(2)
      expect(node.leftHeight).toBe(1)
      expect(node.rightHeight).toBe(2)
    })
  })

  describe('#balanceFactor', () => {
    test('should return tree balance factor', () => {
      let node = new BinaryTreeNode('root')
      node.left = new BinaryTreeNode('left')
      node.right = new BinaryTreeNode('right')
      node.right.right = new BinaryTreeNode('rightright')
      node.right.left = new BinaryTreeNode('rightleft')

      expect(node.balanceFactor).toBe(-1)
      expect(node.right.balanceFactor).toBe(0)
    })
  })

  describe('#parent', () => {
    test('should return parent node', () => {
      let node = new BinaryTreeNode('root')
      node.left = new BinaryTreeNode('left')
      node.right = new BinaryTreeNode('right')
      node.right.right = new BinaryTreeNode('rightright')
      node.right.left = new BinaryTreeNode('rightleft')

      expect(node.parent).toBeNull()
      expect(node.left.parent).toBe(node)
      expect(node.right.parent).toBe(node)
      expect(node.right.left.parent).toBe(node.right)
      expect(node.right.right.parent).toBe(node.right)
    })
  })

  describe('#equals()', () => {
    test('should return whether two node values are equal', () => {
      expect(
        new BinaryTreeNode('root').equals(new BinaryTreeNode('root'))
      ).toBe(true)
      expect(new BinaryTreeNode('foo').equals(new BinaryTreeNode('bar'))).toBe(
        false
      )
    })
  })
})
