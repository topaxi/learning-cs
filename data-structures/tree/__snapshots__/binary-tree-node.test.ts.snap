// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinaryTreeNode<T> #toJSON() should serialize to JSON 1`] = `
Object {
  "left": Object {
    "left": null,
    "right": null,
    "value": "left",
  },
  "right": Object {
    "left": Object {
      "left": null,
      "right": null,
      "value": "rightleft",
    },
    "right": Object {
      "left": null,
      "right": null,
      "value": "rightright",
    },
    "value": "right",
  },
  "value": "root",
}
`;
